### PR TITLE
Simple reduce pipeline

### LINKED ISSUE #289: integrate reduce vertex with RL and PnF

### PR DESCRIPTION
**Pull request by yhl25**
Signed-off-by: Yashash H L <yashashhl25@gmail.com>

Fixes #289

simple pipeline spec with reduce vertex

### CODE REVIEW

File: examples/6-simple-reduce-pipeline.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    - name: even-or-odd
+      udf:
+        container:
+          # Tell the input number is even or odd, see https://github.com/numaproj/numaflow-go/tree/main/pkg/function/examples/evenodd
+          image: quay.io/numaio/go-even-odd-example
+    - name: compute-sum
+      udf:
+        container:
+          # compute the sum
+          image: quay.io/yhl25/reduce-image:master
```
Comment (KeranYang) [inline examples/6-simple-reduce-pipeline.yaml:19]: nit: should we use the `/numaio/` path and add a reference to the source of of the image? (same as the go-even-odd-example above.)

Comment (vigith) [reply examples/6-simple-reduce-pipeline.yaml:19]: +1, I have sent an invite to @yhl25 to join quay.io/numaio.

Comment (yhl25) [reply examples/6-simple-reduce-pipeline.yaml:19]: yes, will do that. I didn't have access o **numaio** so I was using my personal repo.

File: pkg/reduce/readloop/readloop.go [hunk 2]
```diff
@@ -129,28 +129,41 @@ func (rl *ReadLoop) Process(ctx context.Context, messages []*isb.ReadMessage) {
...
-			writeFn := func(ctx context.Context, m *isb.ReadMessage) error {
-				return q.Write(ctx, m)
-			}
-			ctxClosedErr = rl.executeWithBackOff(ctx, writeFn, "failed to Write Message", pbqWriteBackoff, m, partitionID)
+			attempt := 0
+			ctxClosedErr = wait.ExponentialBackoffWithContext(ctx, pbqWriteBackoff, func() (done bool, err error) {
+				rErr := q.Write(ctx, m)
+				attempt += 1
+				if rErr != nil {
+					rl.log.Errorw("failed to write message", zap.Any("msgOffSet", m.ReadOffset.String()), zap.Any("partitionID", partitionID.String()), zap.Any("attempt", attempt), zap.Error(rErr))
```
Comment (whynowy) [inline pkg/reduce/readloop/readloop.go:137]: Capital the first letter in the log message.

File: pkg/reduce/readloop/readloop.go [hunk 3]
```diff
@@ -235,6 +232,12 @@ func (rl *ReadLoop) ShutDown(ctx context.Context) {
 // upsertWindowsAndKeys will create or assigns (if already present) a window to the message. It is an upsert operation
 // because windows are created out of order, but they will be closed in-order.
 func (rl *ReadLoop) upsertWindowsAndKeys(m *isb.ReadMessage) []*keyed.KeyedWindow {
+	// drop the late messages
+	if m.IsLate {
+		rl.log.Infow("Dropping the late message", zap.Time("eventTime", m.EventTime))
```
Comment (vigith) [inline pkg/reduce/readloop/readloop.go:237]:
```suggestion
		rl.log.Warnw("Dropping the late message", zap.Time("eventTime", m.EventTime))
```

**Pull request merged**